components/application_manager/opapp.cpp

```diff
@@ -0,0 +1,131 @@
+#include "opapp.h"
+#include "log.h"
+
+#define COUNT_DOWN_TIMEOUT 60000
+
+static std::string opAppStateToString(const App::E_APP_STATE &state);
+
+/**
+ * Create opapp from url.
+ * 
+ * @throws std::runtime_error
+ */
+OpApp::OpApp(const std::string &url, std::shared_ptr<OpApp::SessionCallback> sessionCallback)
+    : App(url, sessionCallback)
+{ 
+    m_state = BACKGROUND_STATE; // ETSI TS 103 606 V1.2.1 (2024-03) page 36
+}
+
+/**
+ * Create opapp from Ait description.
+ * 
+ * @throws std::runtime_error
+ */
+OpApp::OpApp(const Ait::S_AIT_APP_DESC &desc, bool isNetworkAvailable, std::shared_ptr<OpApp::SessionCallback> sessionCallback)
+    : App(
+        desc,
+        Utils::MakeInvalidDvbTriplet(),
+        isNetworkAvailable,
+        "",
+        true,
+        false,
+        sessionCallback
+    )
+{ 
+    m_state = BACKGROUND_STATE; // ETSI TS 103 606 V1.2.1 (2024-03) page 36
+}
+
+/**
+ * Create opapp from url and inherit another opapp's state (ETSI TS 103 606 V1.2.1 (2024-03) 6.3.3.1).
+ * 
+ * @throws std::runtime_error
+ */
+OpApp::OpApp(const OpApp &other, const std::string &url)
+    : OpApp(url, other.m_sessionCallback)
+{
+    m_state = other.GetState();
+    if (!other.m_countdown.isStopped())
+    {
+        m_countdown.start(other.m_countdown.remaining());
+    }
+}
+
+void OpApp::SetState(const E_APP_STATE &state)
+{
+    if (state != OVERLAID && CanTransitionToState(state))
```

Comment:
Any reason not to include OVERLAID test within CanTransitionToState?

Reply:
Your suggestion changes the logic and the intention of the statement. If state == OVERLAID, the condition fails immediately (no further checks). If state != OVERLAID, it proceeds to check CanTransitionToState(state).
We don't want to check for the OVERLAID state, as it cannot stand as a state on its own, it has to be combined either with FOREGROUND_STATE or TRANSIENT_STATE.

Reply:
This is actually sanity check, not a state where an OpApp can transition to. However, I think it would be better to put it inside of CanTransitionToState() method, and in that case it would return false.